tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -257,4 +260,211 @@ impl ServerKey {
+            let ct = lhs
+                .blocks
+                .iter()
+                .zip(rhs.blocks.iter())
+                .map(|(lhs_block, rhs_block)| self.key.unchecked_sub(lhs_block, rhs_block))
+                .collect::<Vec<_>>();
+            let mut ct = RadixCiphertext::from(ct);
+
+            let generates_or_propagates = self.generate_init_borrow_array(&ct);
+            let (input_borrows, mut output_borrow) =
+                self.compute_borrow_propagation_parallelized_low_latency(generates_or_propagates);
+
+            ct.blocks
+                .par_iter_mut()
+                .zip(input_borrows.par_iter())
+                .for_each(|(block, input_borrow)| {
+                    // Do a true lwe subtraction, as unchecked_sub will adds a correcting term
+                    // to avoid overflow (and trashing padding bit). Here we know each
+                    // block in the ciphertext is >= 1, and that input borrow is either 0 or 1
+                    // so no overflow possible.
+                    crate::core_crypto::algorithms::lwe_ciphertext_sub_assign(
+                        &mut block.ct,
+                        &input_borrow.ct,
+                    );
+                    self.key.message_extract_assign(block);
+                });
+            assert!(ct.block_carries_are_empty());
+            // we know here that the result is a boolean value
+            // however the lut used has a degree of 2.
+            output_borrow.degree.0 = 1;
+            //assert!(output_borrow.degree.0 == 1);
+            (ct, output_borrow)
+        } else {
+            self.unsigned_overflowing_sub(lhs, rhs)
+        }
+    }
+
+    pub(super) fn generate_init_borrow_array(&self, sum_ct: &RadixCiphertext) -> Vec<Ciphertext> {
+        let modulus = self.key.message_modulus.0 as u64;
+
+        // This used for the first pair of blocks
+        // as this pair can either generate or not, but never propagate
+        let lut_does_block_generate_carry = self.key.generate_lookup_table(|x| {
+            if x < modulus {
+                1 // Borrows
+            } else {
+                0 // Does not borrow
+            }
+        });
+
+        let lut_does_block_generate_or_propagate = self.key.generate_lookup_table(|x| {
+            match x.cmp(&modulus) {
+                Ordering::Less => {
+                    1 // Generates borrow
+                }
+                Ordering::Equal => {
+                    2 // Propagates borrow
+                }
+                Ordering::Greater => {
+                    0 // No borrow generated or propagated
+                }
+            }
+        });
+
+        let mut generates_or_propagates = Vec::with_capacity(sum_ct.blocks.len());
+        sum_ct
+            .blocks
+            .par_iter()
+            .enumerate()
+            .map(|(i, block)| {
+                if i == 0 {
+                    // The first block can only output a borrow
```

Comment:
is borrow "just" a carry in the end ? if so maybe changing the name is better for consistency with function calls ? unless it's the standardized term for the subtraction

Reply:
Well, a borrow is the opposite of a carry, you give a carry to the next block, you borrw from the next block

Reply:
And its a standard name borrow == retenue

Reply:
ok thanks !